src/main/java/org/checkerframework/checker/crypto/CryptoAnnotatedTypeFactory.java

```diff
@@ -74,12 +77,83 @@ public boolean isSubtype(final AnnotationMirror subtype, final AnnotationMirror
 
         @Override
         public AnnotationMirror greatestLowerBound(AnnotationMirror a1, AnnotationMirror a2) {
-            return CryptoAnnotatedTypeFactory.this.BOTTOM;
+            if (AnnotationUtils.areSameByName(a1, BOTTOM)
```

Comment:
The `CryptoQualifierHierarchy` currently extends the `ElementQualifierHierarchy`.
Have you tried using https://github.com/eisop/checker-framework/blob/master/framework/src/main/java/org/checkerframework/framework/type/SubtypeIsSubsetQualifierHierarchy.java (or https://github.com/eisop/checker-framework/blob/master/framework/src/main/java/org/checkerframework/framework/type/SubtypeIsSupersetQualifierHierarchy.java in case the relationship is the other way around)?

Reply:
It seems both of them cannot handle the situation where one is AllowedAlgorithms and the other is AllowedProviders, as line 75 shows it will always throw an error..So I do not change it to these two. Maybe we can talk out it tomorrow.

Reply:
Are you sure you are hitting that exception?
The usage of that method is

````
        if (lub.hasElements()) {
            return leastUpperBoundWithElements(a1, qual1, a2, qual2, lub);
        }
````

For you, when AllowedProviders and AllowedAlgorithms is compared, the LUB will be UNKNOWNALGORITHMORPROVIDER, which doesn't have elements, so `leastUpperBoundWithElements` shouldn't be called and there shouldn't be an exception.

Even if it had elements, you would just need to override that method and still share much more of the logic with one of these as superclass.

Reply:
Yeah, it does work.